std/typecons.d

```diff
@@ -3431,6 +3431,32 @@ Params:
     assert(!npi.isNull);
 }
 
+    /**
+      If they are both null, then they are equal. If one is null and the other
+      is not, then they are not equal. If they are both non-null, then they are
+      equal if their values are equal.
+      */
+    bool opEquals()(auto ref const(typeof(this)) rhs) const
+    {
+        return _value == rhs._value;
```

Comment:
Ugh, I hate to do this to you, but actually there needs to be one static if: if `_value` is a floating point, then this won't work, but `isNull` will. So you need to duplicate what `isNull` does in that instance.

Reply:
We can just do:
```d
bool opEquals()(auto ref const(typeof(this)) rhs) const
{
    // We check __traits(compiles) in case opEquals is not CTFE-able.
    static if (__traits(compiles, { static assert(nullValue == nullValue); }))
        return _value == rhs._value;
    else
        // Note that it is wrong to use this branch for floating point T
        // when nullValue is not NaN.
        return _value is rhs._value;
}
```

Reply:
@n8sh No. That else block would malfunction if `typeof(_value)` is a class with custom opEquals and neither of the operands are null. Otherwise, no strict opinions on this.